### PR TITLE
Support petersburg in evmVersion

### LINKED ISSUE #6086: EVM version: "petersburg"
Constantinople was launched on the testnets, but was cancelled last minute on mainnet. It is to be launched on the same block as "Petersburg", which disables "net gas metering" introduced in Constantinople.

Since we do not support net gas metering in the gas estimator this should have no effect for us.

While it feels strange to have two hard forks, which in practice mean a single one, I still suggest we introduced "petersburg" as an entry in our EVMVersion listed to avoid any further confusion with users.

It is not fully clear if the community will refer to this hardfork as "constantinople" or "petersburg", but technically all documentation and clients are referring to it separately.

### PR DESCRIPTION
**Pull request by axic**
Closes #6086.

Ref: https://blog.ethereum.org/2019/02/22/ethereum-constantinople-st-petersburg-upgrade-announcement/

### CODE REVIEW

File: liblangutil/EVMVersion.h [hunk 1]
```diff
@@ -30,7 +30,7 @@ namespace langutil
 
 /**
  * A version specifier of the EVM we want to compile to.
- * Defaults to the latest version.
+ * Defaults to the latest version deployed on Ethereum mainnet.
```
Comment (chriseth) [inline liblangutil/EVMVersion.h]: At the time of compiler release.

File: liblangutil/EVMVersion.h [hunk 2]
```diff
@@ -81,7 +83,7 @@ class EVMVersion:
 	bool canOverchargeGasForCall() const { return *this >= tangerineWhistle(); }
 
 private:
-	enum class Version { Homestead, TangerineWhistle, SpuriousDragon, Byzantium, Constantinople };
+	enum class Version { Homestead, TangerineWhistle, SpuriousDragon, Byzantium, Constantinople, Petersburg };
 
 	EVMVersion(Version _version): m_version(_version) {}
```
Comment (chriseth) [inline liblangutil/EVMVersion.h:89]: The line below needs to be updated - or do you want do do that in a separate PR?

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 3]
```diff
@@ -325,7 +325,7 @@ bool ReferencesResolver::visit(InlineAssembly const& _inlineAssembly)
 		analysisInfo,
 		errorsIgnored,
 		errorTypeForLoose,
-		yul::EVMDialect::looseAssemblyForEVM(EVMVersion::constantinople()),
+		yul::EVMDialect::looseAssemblyForEVM(EVMVersion::petersburg()),
```
Comment (chriseth) [inline libsolidity/analysis/ReferencesResolver.cpp:328]: Shouldn't this just use the default constructor instead?

File: libsolidity/parsing/Parser.cpp [hunk 4]
```diff
@@ -1042,7 +1042,7 @@ ASTPointer<InlineAssembly> Parser::parseInlineAssembly(ASTPointer<ASTString> con
 	}
 
 	// Using latest EVM Version for now, it will be run again later.
-	yul::Parser asmParser(m_errorReporter, yul::EVMDialect::looseAssemblyForEVM(EVMVersion::constantinople()));
+	yul::Parser asmParser(m_errorReporter, yul::EVMDialect::looseAssemblyForEVM(EVMVersion::petersburg()));
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:1045]: Same here - otherwise we need to update a lot of places.

File: test/tools/yulopti.cpp [hunk 5]
```diff
@@ -206,7 +206,7 @@ class YulOpti
 private:
 	ErrorList m_errors;
 	shared_ptr<yul::Block> m_ast;
-	shared_ptr<Dialect> m_dialect{EVMDialect::strictAssemblyForEVMObjects(EVMVersion::constantinople())};
+	shared_ptr<Dialect> m_dialect{EVMDialect::strictAssemblyForEVMObjects(EVMVersion::petersburg())};
```
Comment (chriseth) [inline test/tools/yulopti.cpp:209]: Default constructor?

File: liblangutil/EVMVersion.h (hunk 2 above)
Comment (axic) [reply liblangutil/EVMVersion.h:89]: The default version? Yes, I think that belongs in a separate one as it changes other parts of the code/tests.

File: libsolidity/analysis/ReferencesResolver.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/analysis/ReferencesResolver.cpp:328]: I think #5959 also touches on these so would prefer to merge that first.

**Comment by axic** [comment]
Need to also include documentation.

**Comment by chriseth** [comment]
Taking this over.

**Comment by axic** [comment]
> Taking this over.

Lets merge #5959 first, I can finish this after.

File: libsolidity/analysis/ReferencesResolver.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/analysis/ReferencesResolver.cpp:328]: Ah I see, this just updates all references to `constantinople` by the more or less identical `petersburg`.

**Comment by chriseth** [comment]
Not sure what to document, but I think this is fine to merge.

**Comment by chriseth** [comment]
https://github.com/ethereum/solidity/pull/5959 is ready for final review.

**Comment by chriseth** [comment]
@axic not that this all has to happen in the next 3-4 hours.

**Comment by chriseth** [comment]
Missing a change in CommandlineInterface.cpp.

**Comment by axic** [comment]
> next 3-4 hours.

Isnt release tomorrow and not today?

**Comment by chriseth** [comment]
Yes, it is tomorrow, so the code has to be finished today, also to have a nightly binary that can be tested in remix.

**Comment by axic** [comment]
Even if there's no conflict I'd rebase this due to the optimiser changes.

File: docs/using-the-compiler.rst [hunk 6]
```diff
@@ -108,18 +108,22 @@ at each version. Backward compatibility is not guaranteed between each version.
...
+   - The ``staticcall`` opcode is used when calling non-library view or pure functions, which prevents the functions from modifying state at the EVM level, i.e., even applies when you use invalid type conversions.
+   - It is possible to access dynamic data returned from function calls.
    - ``revert`` opcode introduced, which means that ``revert()`` will not waste gas.
-- ``constantinople`` (still in progress)
-   - opcodes ``shl``, ``shr`` and ``sar`` are available in assembly.
-   - shifting operators use shifting opcodes and thus need less gas.
+- ``constantinople``
+   - Opcodes ``shl``, ``shr`` and ``sar`` are available in assembly.
+   - Shifting operators use shifting opcodes and thus need less gas.
+   - Opcodes ``create2`` and ``extcodehash`` are available in assembly.
```
Comment (axic) [inline docs/using-the-compiler.rst]: Why is this separate from the shift list?

File: docs/using-the-compiler.rst [hunk 7]
```diff
@@ -108,18 +108,22 @@ at each version. Backward compatibility is not guaranteed between each version.
...
+- ``petersburg``
+   - identical to constantinople
```
Comment (axic) [inline docs/using-the-compiler.rst]: Why isnt this capitalised if the others are?

File: docs/using-the-compiler.rst (hunk 6 above)
Comment (chriseth) [reply docs/using-the-compiler.rst]: Because it is conceptually different. I can merge the lists if you like.

File: docs/using-the-compiler.rst (hunk 7 above)
Comment (chriseth) [reply docs/using-the-compiler.rst]: Because it is not a full sentence.

File: docs/using-the-compiler.rst (hunk 6 above)
Comment (axic) [reply docs/using-the-compiler.rst]: The other versions don't have that separation. On that note, shouldn't create2 / extcodehash also separated then?

**Comment by chriseth** [comment]
Rebased and updated commandline compiler.

File: docs/using-the-compiler.rst (hunk 7 above)
Comment (axic) [reply docs/using-the-compiler.rst]: Can you make it into a sentence (e.g. have a dot?) :)

**Comment by chriseth** [comment]
Updated.

**Comment by axic** [comment]
Looks good, cannot approve myself.
